File: test/domain/insights/data_confidence_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openvitals/core/period/time_range.dart';
import 'package:openvitals/core/time/local_date.dart';
import 'package:openvitals/domain/insights/data_confidence.dart';

void main() {
  test('confidence calculates coverage inside the selected period', () {
    final period = DatePeriod(LocalDate(2026, 5, 1), LocalDate(2026, 5, 7));

    final confidence = dataConfidence(
      period,
      [
        LocalDate(2026, 4, 30),
        LocalDate(2026, 5, 1),
        LocalDate(2026, 5, 3),
        LocalDate(2026, 5, 3),
        LocalDate(2026, 5, 8),
      ],
      5,
      sources: const ['com.example.watch'],
    );

    expect(confidence.expectedDays, 7);
    expect(confidence.trackedDays, 2);
    expect(confidence.coveragePercent, 29);
  });

  test('mixed sources are reported as medium confidence', () {
    final confidence = dataConfidence(
      DatePeriod(LocalDate(2026, 5, 1), LocalDate(2026, 5, 1)),
      [LocalDate(2026, 5, 1)],
      12,
      sources: const ['com.example.watch', 'com.example.scale'],
    );

    expect(confidence.level, DataConfidenceLevel.medium);
    expect(confidence.sourceConsistency, DataSourceConsistency.mixedSources);
    expect(confidence.warnings, contains(DataConfidenceWarning.mixedSources));
  });

  test('calculated values include a calculated warning', () {
    final confidence = dataConfidence(
      DatePeriod(LocalDate(2026, 5, 1), LocalDate(2026, 5, 1)),
      [LocalDate(2026, 5, 1)],
      1,
      sources: const ['com.example.scale'],
      valueKind: DataValueKind.calculated,
    );

    expect(confidence.level, DataConfidenceLevel.low);
    expect(confidence.warnings, contains(DataConfidenceWarning.calculatedValue));
  });

  test('empty samples are low confidence', () {
    final confidence = dataConfidence(
      DatePeriod(LocalDate(2026, 5, 1), LocalDate(2026, 5, 31)),
      const [],
      0,
    );

    expect(confidence.level, DataConfidenceLevel.low);
    expect(confidence.warnings, contains(DataConfidenceWarning.noSourceDetails));
  });

  test('manual entries are reported', () {
    final confidence = dataConfidence(
      DatePeriod(LocalDate(2026, 5, 1), LocalDate(2026, 5, 3)),
      [
        LocalDate(2026, 5, 1),
        LocalDate(2026, 5, 2),
        LocalDate(2026, 5, 3),
      ],
      3,
      sources: const ['com.example.app'],
      manualEntryCount: 1,
    );

    expect(confidence.level, DataConfidenceLevel.medium);
    expect(confidence.warnings, contains(DataConfidenceWarning.manualEntries));
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/domain/insights/DataConfidenceTest.kt
package tech.mmarca.openvitals.domain.insights

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import tech.mmarca.openvitals.core.period.DatePeriod
import java.time.LocalDate

class DataConfidenceTest {

    @Test fun `confidence calculates coverage inside the selected period`() {
        val period = DatePeriod(LocalDate.of(2026, 5, 1), LocalDate.of(2026, 5, 7))

        val confidence = dataConfidence(
            period = period,
            trackedDates = listOf(
                LocalDate.of(2026, 4, 30),
                LocalDate.of(2026, 5, 1),
                LocalDate.of(2026, 5, 3),
                LocalDate.of(2026, 5, 3),
                LocalDate.of(2026, 5, 8),
            ),
            sampleCount = 5,
            sources = listOf("com.example.watch"),
        )

        assertEquals(7, confidence.expectedDays)
        assertEquals(2, confidence.trackedDays)
        assertEquals(29, confidence.coveragePercent)
    }

    @Test fun `mixed sources are reported as medium confidence`() {
        val confidence = dataConfidence(
            period = DatePeriod(LocalDate.of(2026, 5, 1), LocalDate.of(2026, 5, 1)),
            trackedDates = listOf(LocalDate.of(2026, 5, 1)),
            sampleCount = 12,
            sources = listOf("com.example.watch", "com.example.scale"),
        )

        assertEquals(DataConfidenceLevel.MEDIUM, confidence.level)
        assertEquals(DataSourceConsistency.MIXED_SOURCES, confidence.sourceConsistency)
        assertTrue(confidence.warnings.contains(DataConfidenceWarning.MIXED_SOURCES))
    }

    @Test fun `calculated values include a calculated warning`() {
        val confidence = dataConfidence(
            period = DatePeriod(LocalDate.of(2026, 5, 1), LocalDate.of(2026, 5, 1)),
            trackedDates = listOf(LocalDate.of(2026, 5, 1)),
            sampleCount = 1,
            sources = listOf("com.example.scale"),
            valueKind = DataValueKind.CALCULATED,
        )

        assertEquals(DataConfidenceLevel.LOW, confidence.level)
        assertTrue(confidence.warnings.contains(DataConfidenceWarning.CALCULATED_VALUE))
    }

    @Test fun `empty samples are low confidence`() {
        val confidence = dataConfidence(
            period = DatePeriod(LocalDate.of(2026, 5, 1), LocalDate.of(2026, 5, 31)),
            trackedDates = emptyList(),
            sampleCount = 0,
        )

        assertEquals(DataConfidenceLevel.LOW, confidence.level)
        assertTrue(confidence.warnings.contains(DataConfidenceWarning.NO_SOURCE_DETAILS))
    }

    @Test fun `manual entries are reported`() {
        val confidence = dataConfidence(
            period = DatePeriod(LocalDate.of(2026, 5, 1), LocalDate.of(2026, 5, 3)),
            trackedDates = listOf(
                LocalDate.of(2026, 5, 1),
                LocalDate.of(2026, 5, 2),
                LocalDate.of(2026, 5, 3),
            ),
            sampleCount = 3,
            sources = listOf("com.example.app"),
            manualEntryCount = 1,
        )

        assertEquals(DataConfidenceLevel.MEDIUM, confidence.level)
        assertTrue(confidence.warnings.contains(DataConfidenceWarning.MANUAL_ENTRIES))
    }
}
